washingtonpost/FastFEC
# Cross-compiled Python package/distribution

## Description

This PR adds a script to cross-compile wheels for the FastFEC Python package along with a GitHub actions workflow to generate wheels for all relevant OS's. This approach subverts the typical `setup.py` script in favor of a `make_wheels.py` script that automatically constructs the wheels for each OS that is based on https://github.com/ziglang/zig-pypi/blob/main/make_wheels.py

Note that there are many commits in this PR. I tried to get it working with [cibuildwheel](https://cibuildwheel.readthedocs.io/en/stable/) first, which would be the de facto way to do cross-platform builds, but each build took over 30 minutes and there were OS-specific issues that were hard to debug. This approach mirrors the way the actual FastFEC package is built and should be just as stable.

Also if for whatever reason the wheel does not work, `setup.py` will automatically run as backup when `pip install fastfec` happens in the future.

To verify this PR works, I launched a Windows VM and confirmed that the Python library could be installed. I also launched AWS Linux x86_64 and aarch64 (Gravitron) instances and confirmed both could install the Python wheel.

## Jira ticket

https://arcpublishing.atlassian.net/browse/ELEX-141

## Test steps

* Go here https://github.com/washingtonpost/FastFEC/actions/runs/1668364069
* Click the artifact file to download it
* Extract the zip contents of the artifact
* Find the path to the `.whl` file inside that has your desired architecture
* Run `pip install {path to the the desired whl file}`
* Run the attached `test.py` and it should print JSON-esque output and not error!

test.py:
```python
from fastfec import FastFEC
from io import BytesIO

file = b""""HDR","FEC","5.1","Navision AVF","3.00","^","","000",""
"F3XN","C00397000","South Dakota Women Vote!","1120 Connecticut Avenue NW","Ste 1100","Washington","DC","20036","","","TER","","","","20041001","20041122","61571.00","0.00","61571.00","61571.00","0.00","0.00","0.00","0.00","0.00","0.00","0.00","0.00","0.00","0.00","0.00","0.00","0.00","0.00","0.00","0.00","0.00","0.00","0.00","0.00","0.00","0.00","61571.00","0.00","0.00","0.00","0.00","0.00","0.00","0.00","0.00","0.00","0.00","61571.00","61571.00","0.00","0.00","0.00","0.00","0.00","0.00","0.00","2004","61775.00","61775.00","61775.00","0.00","61750.00","25.00","61775.00","0.00","0.00","61775.00","0.00","0.00","0.00","0.00","0.00","0.00","0.00","61775.00","61775.00","0.00","0.00","204.00","204.00","61571.00","0.00","0.00","0.00","0.00","0.00","0.00","0.00","0.00","0.00","0.00","61775.00","61775.00","61775.00","0.00","61775.00","204.00","0.00","204.00","Caroline C. Fines","20041201","0.00","0.00","0.00","0.00","0.00","0.00","0.00","0.00","0.00","0.00","0.00","0.00"
"SB22","C00397000","PTY","","207 East Capitol #103","PO Box 737","Pierre","SD","57501","","Transfer to Affiliate","","","20041019","61144.66","","","","","","","","","","","","","","","","SB22-2533","","","","","","","South Dakota Democratic Party ","","","","",""
"SB22","C00397000","PAC","","805 15th St., NW #400","","Washington","DC","20005","","Transfer to Affiliate","","","20041019","426.34","","","","","","","","","","","","","","","","SB22-2532","","","","","","","EMILY's List Federal Operating ","","","","",""
"""

with BytesIO(file) as f:
    with FastFEC() as fastfec:
        for form, line in fastfec.parse(f):
            print("GOT", form, line)
```
